In [2]:
import pandas as pd
import numpy as np

In [39]:
cols = [ 'CONTRACTING_AGENCY_ID', 
        'CONTRACTING_AGENCY_NAME', 
        'FUNDING_AGENCY_ID',
        'FUNDING_AGENCY_NAME',
        'AGENCY_CODE',
        'PIID_AGENCY_NAME',
        'ORIGINAL_AGENCY_CODE',
        'REFERENCE_AGENCY_CODE',
        'FEDERAL_AGENCY',
        'AGENCY',
        'AGENCY_IDENTIFIER',
        'IDV_CONTRACTING_AGENCY_ID',
        'IDV_CONTRACTING_AGENCY_NAME',
        'IDV_REF_IDV_AGENCY_CODE',
        'CLOSED_BY']

In [3]:
cd /Users/karolina/Desktop/SBA

/Users/karolina/Desktop/SBA


In [40]:
test_2022 = pd.read_parquet('2022.parquet', engine='pyarrow',columns=cols)

In [5]:
test_2022

CONTRACTING_AGENCY_ID           CONTRACTING_AGENCY_NAME  \
0                        97AS          DEFENSE LOGISTICS AGENCY   
1                        97AS          DEFENSE LOGISTICS AGENCY   
2                        97AS          DEFENSE LOGISTICS AGENCY   
3                        97AS          DEFENSE LOGISTICS AGENCY   
4                        97AS          DEFENSE LOGISTICS AGENCY   
...                       ...                               ...   
6088449                  97AS          DEFENSE LOGISTICS AGENCY   
6088450                  2100                  DEPT OF THE ARMY   
6088451                  4732       FEDERAL ACQUISITION SERVICE   
6088452                  7008                  U.S. COAST GUARD   
6088453                  9543  MILLENNIUM CHALLENGE CORPORATION   

        FUNDING_AGENCY_ID               FUNDING_AGENCY_NAME AGENCY_CODE  \
0                    97AS          DEFENSE LOGISTICS AGENCY        9700   
1                    97AS          DEFENSE LOGISTICS AGENCY        9700   
2                    97AS          DEFENSE LOGISTICS AGENCY        9700   
3                    97AS          DEFENSE LOGISTICS AGENCY        9700   
4                    97AS          DEFENSE LOGISTICS AGENCY        9700   
...                   ...                               ...         ...   
6088449              97AS          DEFENSE LOGISTICS AGENCY        9700   
6088450              2100                  DEPT OF THE ARMY        9700   
6088451              4732       FEDERAL ACQUISITION SERVICE        4732   
6088452              7008                  U.S. COAST GUARD        7008   
6088453              9543  MILLENNIUM CHALLENGE CORPORATION        9543   

                         PIID_AGENCY_NAME ORIGINAL_AGENCY_CODE  \
0                         DEPT OF DEFENSE                 None   
1                         DEPT OF DEFENSE                 None   
2                         DEPT OF DEFENSE                 None   
3                         DEPT OF DEFENSE                 None   
4                         DEPT OF DEFENSE                 None   
...                                   ...                  ...   
6088449                   DEPT OF DEFENSE                 None   
6088450                   DEPT OF DEFENSE                 None   
6088451       FEDERAL ACQUISITION SERVICE                 None   
6088452                  U.S. COAST GUARD                 None   
6088453  MILLENNIUM CHALLENGE CORPORATION                 None   

        REFERENCE_AGENCY_CODE FEDERAL_AGENCY    AGENCY AGENCY_IDENTIFIER  \
0                        9700             NO  ORG-2116              None   
1                        9700             NO  ORG-2116              None   
2                        9700             NO  ORG-2116              None   
3                        9700             NO  ORG-2116              None   
4                        9700             NO  ORG-2116              None   
...                       ...            ...       ...               ...   
6088449                  9700             NO  ORG-2116              None   
6088450                  None             NO  ORG-2398                21   
6088451                  4732             NO  ORG-4530              None   
6088452                  7008             NO  ORG-2367              None   
6088453                  None             NO  ORG-1412                95   

        IDV_CONTRACTING_AGENCY_ID  IDV_CONTRACTING_AGENCY_NAME  \
0                            97AS     DEFENSE LOGISTICS AGENCY   
1                            97AS     DEFENSE LOGISTICS AGENCY   
2                            97AS     DEFENSE LOGISTICS AGENCY   
3                            97AS     DEFENSE LOGISTICS AGENCY   
4                            97AS     DEFENSE LOGISTICS AGENCY   
...                           ...                          ...   
6088449                      97AS     DEFENSE LOGISTICS AGENCY   
6088450                      None                         None   
60

In [6]:
test_2022.isna().sum()

CONTRACTING_AGENCY_ID                0
CONTRACTING_AGENCY_NAME              0
FUNDING_AGENCY_ID                    0
FUNDING_AGENCY_NAME                  0
AGENCY_CODE                          0
PIID_AGENCY_NAME                     0
ORIGINAL_AGENCY_CODE           6088454
REFERENCE_AGENCY_CODE          1142128
FEDERAL_AGENCY                       0
AGENCY                               0
AGENCY_IDENTIFIER              5899329
IDV_CONTRACTING_AGENCY_ID      1142172
IDV_CONTRACTING_AGENCY_NAME    1142172
IDV_REF_IDV_AGENCY_CODE        5262532
CLOSED_BY                      3063154
dtype: int64

In [10]:
(test_2022['CONTRACTING_AGENCY_ID']==test_2022['FUNDING_AGENCY_ID']).value_counts()

True     5906289
False     182165
Name: count, dtype: int64

In [11]:
(test_2022['CONTRACTING_AGENCY_NAME']==test_2022['FUNDING_AGENCY_NAME']).value_counts()

True     5906289
False     182165
Name: count, dtype: int64

In [12]:
(test_2022['AGENCY_CODE']==test_2022['REFERENCE_AGENCY_CODE']).value_counts()

True     4733716
False    1354738
Name: count, dtype: int64

In [14]:
(test_2022['CONTRACTING_AGENCY_NAME']==test_2022['PIID_AGENCY_NAME']).value_counts()

False    4135313
True     1953141
Name: count, dtype: int64

In [27]:
test_2022['PIID_AGENCY_NAME'].value_counts()

PIID_AGENCY_NAME
DEPT OF DEFENSE                              4133421
FEDERAL ACQUISITION SERVICE                  1206172
VETERANS AFFAIRS, DEPARTMENT OF               129060
STATE, DEPARTMENT OF                           67487
FEDERAL PRISON SYSTEM / BUREAU OF PRISONS      43349
                                              ...   
HEALTH AND HUMAN SERVICES, DEPARTMENT OF           2
FEDERAL HOUSING FINANCE AGENCY                     2
COMMERCE, DEPARTMENT OF                            1
TECHNOLOGY TRANSFORMATION SERVICE                  1
TRANSPORTATION, DEPARTMENT OF                      1
Name: count, Length: 143, dtype: int64

In [28]:
test_2022['CONTRACTING_AGENCY_NAME'].value_counts()

CONTRACTING_AGENCY_NAME
DEFENSE LOGISTICS AGENCY                          3613050
FEDERAL ACQUISITION SERVICE                       1206172
DEPT OF THE NAVY                                   177759
DEPT OF THE ARMY                                   135390
VETERANS AFFAIRS, DEPARTMENT OF                    129062
                                                   ...   
INSPECTOR GENERAL, OFFICE OF                            8
ADMINISTRATIVE CONFERENCE OF THE U. S.                  7
CHEMICAL SAFETY AND HAZARD INVESTIGATION BOARD          7
FEDERAL HOUSING FINANCE AGENCY                          2
TECHNOLOGY TRANSFORMATION SERVICE                       1
Name: count, Length: 161, dtype: int64

In [23]:
(test_2022['IDV_CONTRACTING_AGENCY_ID']==test_2022['CONTRACTING_AGENCY_ID']).value_counts()

True     4716606
False    1371848
Name: count, dtype: int64

In [21]:
test_2022['FEDERAL_AGENCY'].value_counts()

FEDERAL_AGENCY
NO     6083912
YES       4542
Name: count, dtype: int64

In [22]:
test_2022['AGENCY'].value_counts()

AGENCY
ORG-2116    3613050
ORG-4530    1206172
ORG-2410     177759
ORG-2398     135390
ORG-1408     129062
             ...   
ORG-2123          8
ORG-1338          7
ORG-1415          7
ORG-4327          2
ORG-5502          1
Name: count, Length: 168, dtype: int64

In [24]:
test_2022['CONTRACTING_AGENCY_ID'].value_counts()

CONTRACTING_AGENCY_ID
97AS    3613050
4732    1206172
1700     177759
2100     135390
3600     129062
         ...   
97EX          8
9515          7
9565          7
9542          2
4775          1
Name: count, Length: 168, dtype: int64

In [25]:
test_2022['AGENCY_CODE'].value_counts()

AGENCY_CODE
9700    4133421
4732    1181648
3600     129060
1900      67487
1540      43349
         ...   
9542          2
5940          1
1300          1
4775          1
6900          1
Name: count, Length: 153, dtype: int64

## Notes

**CONTRACTING_AGENCY_ID**            0 - don't always match with funding agency ID
CONTRACTING_AGENCY_NAME              0
FUNDING_AGENCY_ID                    0
FUNDING_AGENCY_NAME                  0
**AGENCY_CODE**                      0 - different than ID
PIID_AGENCY_NAME                     0 - looks like its department
ORIGINAL_AGENCY_CODE           6088454 - almost all missing
REFERENCE_AGENCY_CODE          1142128 - same as AGENCY_CODE but with missing values 
FEDERAL_AGENCY                       0 - Yes or No 
**AGENCY**                           0 - different ID for agency - but match 
AGENCY_IDENTIFIER              5899329 - almost all missing 
IDV_CONTRACTING_AGENCY_ID      1142172 - mostly same as CONTRACTING_AGENCY_ID
IDV_CONTRACTING_AGENCY_NAME    1142172 - mostly same as CONTRACTING_AGENCY_NAME
IDV_REF_IDV_AGENCY_CODE        5262532 - almosy all missing 
CLOSED_BY                      3063154 - half missing 

In [13]:
test_2022.shape

(6088454, 15)

In [20]:
test_2022['REFERENCE_AGENCY_CODE'].isna().sum()

1142128

### Notes

In [46]:
test_2022[['CONTRACTING_AGENCY_ID','AGENCY_CODE','AGENCY']]

CONTRACTING_AGENCY_ID AGENCY_CODE    AGENCY
0                        97AS        9700  ORG-2116
1                        97AS        9700  ORG-2116
2                        97AS        9700  ORG-2116
3                        97AS        9700  ORG-2116
4                        97AS        9700  ORG-2116
...                       ...         ...       ...
6088449                  97AS        9700  ORG-2116
6088450                  2100        9700  ORG-2398
6088451                  4732        4732  ORG-4530
6088452                  7008        7008  ORG-2367
6088453                  9543        9543  ORG-1412

[6088454 rows x 3 columns]

In [41]:
(test_2022['CONTRACTING_AGENCY_ID']==test_2022['AGENCY_CODE']).value_counts()

False    4159838
True     1928616
Name: count, dtype: int64

In [42]:
(test_2022['CONTRACTING_AGENCY_ID']==test_2022['AGENCY']).value_counts()

False    6088454
Name: count, dtype: int64

In [43]:
(test_2022['AGENCY_CODE']==test_2022['AGENCY']).value_counts()

False    6088454
Name: count, dtype: int64